### ISSUE TITLE
DDL autogeneration doesn't work due to weak table existence detection

### ISSUE THREAD

**Issue by xshaka** [Impetus/kundera#792]
Testing the DDL autogeneration for Cassandra with:

```
        <property name="kundera.ddl.auto.prepare" value="update" />
```

in the persistence.xml, my code fails during the startup when tables already exist. Digging into the code, I have found that the mechanism used to check table existence is particularly weak, since it's based on exception messages. I've found this in class CassandraSchemaManager:

```
private void updateExistingColumnFamily(TableInfo tableInfo, KsDef ksDef, InvalidRequestException irex)
        throws Exception
{
    StringBuilder builder = new StringBuilder("Cannot add already existing column family ");

    if (irex.getWhy() != null && irex.getWhy().contains(builder.toString()))
    {
    ....
```

The problem then, is that the exception 'why' says something like "Cannot add already existing table " which clearly breaks the whole mechanism. Isn't possible to ask for table existence using CQL itself? I understand that there are some queries that might help, for example:

```
SELECT * FROM System.schema_columnfamilies WHERE keyspace_name='<keySpace>' AND columnfamily_name='<table_name>';
```

This works for at least Cassandra 2.2.1 and CQL 3.1, which is my current environment.

**Comment by karthikprasad13** [comment]
@xshaka 

We are looking into it and will update soon.

-Karthik

**Comment by karthikprasad13** [comment]
@xshaka 

We have verified the auto-generation logic and it looks fine. Can you share the error logs or the unexpected behavior that you encountered?

-Karthik

**Comment by xshaka** [comment]
First at all, I'm using Kundera 3.1 via Maven. This is my current persistence.xml:

```
<?xml version="1.0" encoding="UTF-8" ?>
<persistence xmlns="http://xmlns.jcp.org/xml/ns/persistence"
             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
             xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/persistence
http://xmlns.jcp.org/xml/ns/persistence/persistence_2_1.xsd"
             version="2.1">
    <persistence-unit name="portal-demo">
        <provider>com.impetus.kundera.KunderaPersistence</provider>
        <properties>
            <property name="kundera.nodes" value="127.0.0.1"/>
            <property name="kundera.port" value="9042"/>
            <property name="kundera.username" value="shaka"/>
            <property name="kundera.password" value="algorithm"/>
            <property name="kundera.keyspace" value="shaka_test"/>
            <property name="kundera.dialect" value="cassandra"/>
            <property name="kundera.show.query" value="true" />
            <property name="kundera.ddl.auto.prepare" value="update" />
            <property name="kundera.client.lookup.class"
                      value="com.impetus.kundera.client.cassandra.dsdriver.DSClientFactory"/>
        </properties>
    </persistence-unit>
</persistence>
```

With that configuration, during the first running I see these CQL sentences, which means everything goes OK:

```
CREATE KEYSPACE IF NOT EXISTS "shaka_test" WITH REPLICATION = { 'class':'org.apache.cassandra.locator.SimpleStrategy','replication_factor':1} and durable_writes = 'true'
CREATE COLUMNFAMILY "subject" ("subjectName" text ,"subjectDescription" text ,"id" uuid  , PRIMARY KEY("id"))
CREATE COLUMNFAMILY "professor_student_assignment" ("profesorId" uuid ,"studentId" uuid ,"key" text  , PRIMARY KEY("key"))
create index if not exists on "professor_student_assignment"("profesorId")
create index if not exists on "professor_student_assignment"("studentId")
CREATE COLUMNFAMILY "person" ("personName" text ,"address" text ,"birthDate" timestamp ,"type" text ,"taughtSubjectId" uuid ,"collegeId" uuid ,"id" uuid  , PRIMARY KEY("id"))
create index if not exists on "person"("type")
create index if not exists on "person"("taughtSubjectId")
create index if not exists on "person"("collegeId")
CREATE COLUMNFAMILY "qualification" ("mark" double ,"evaluationTimestamp" timestamp ,"subjectId" uuid ,"studentId" uuid ,"id" uuid  , PRIMARY KEY("id"))
create index if not exists on "qualification"("subjectId")
create index if not exists on "qualification"("studentId")
CREATE COLUMNFAMILY "college_subject_mapping" ("collegeId" uuid ,"subjectId" uuid ,"key" text  , PRIMARY KEY("key"))
create index if not exists on "college_subject_mapping"("collegeId")
create index if not exists on "college_subject_mapping"("subjectId")
```

But, at second run, with same model and configuration, I get these:

```
CREATE COLUMNFAMILY "subject" ("subjectDescription" text ,"subjectName" text ,"id" uuid  , PRIMARY KEY("id"))
2015-10-09 10:30:10,360 [main] ERROR com.impetus.client.cassandra.schemamanager.CassandraSchemaManager - Error occurred while creating tablesubject, Caused by: InvalidRequestException(why:Cannot add already existing table "subject" to keyspace "shaka_test").
2015-10-09 10:30:10,360 [main] ERROR com.impetus.client.cassandra.schemamanager.CassandraSchemaManager - Error occurred while creating shaka_test, Caused by: .
com.impetus.kundera.configure.schema.SchemaGenerationException: Error occurred while creating table subject
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.updateExistingColumnFamily(CassandraSchemaManager.java:573)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateUsingCQL3(CassandraSchemaManager.java:820)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createColumnFamilies(CassandraSchemaManager.java:430)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateKeyspace(CassandraSchemaManager.java:279)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.update(CassandraSchemaManager.java:627)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.handleOperations(AbstractSchemaManager.java:267)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.exportSchema(AbstractSchemaManager.java:115)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.exportSchema(CassandraSchemaManager.java:166)
    at com.impetus.kundera.configure.SchemaConfiguration.configure(SchemaConfiguration.java:188)
    at com.impetus.kundera.configure.ClientMetadataBuilder.buildClientFactoryMetadata(ClientMetadataBuilder.java:48)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configureClientFactories(EntityManagerFactoryImpl.java:408)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configure(EntityManagerFactoryImpl.java:161)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.<init>(EntityManagerFactoryImpl.java:116)
    at com.impetus.kundera.KunderaPersistence.createContainerEntityManagerFactory(KunderaPersistence.java:63)
    at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:343)
    at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:318)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
    at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:305)
    at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
    at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:301)
    at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:196)
    at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1045)
    at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:824)
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:537)
    at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:687)
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
    at org.springframework.boot.test.SpringApplicationContextLoader.loadContext(SpringApplicationContextLoader.java:104)
    at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
    at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
    at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
    at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:183)
    at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:123)
    at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:228)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:217)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:276)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:278)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:236)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:87)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
    at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:180)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: InvalidRequestException(why:Cannot add already existing table "subject" to keyspace "shaka_test")
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result$execute_cql3_query_resultStandardScheme.read(Cassandra.java:50297)
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result$execute_cql3_query_resultStandardScheme.read(Cassandra.java:50274)
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result.read(Cassandra.java:50189)
    at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:78)
    at org.apache.cassandra.thrift.Cassandra$Client.recv_execute_cql3_query(Cassandra.java:1734)
    at org.apache.cassandra.thrift.Cassandra$Client.execute_cql3_query(Cassandra.java:1719)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateUsingCQL3(CassandraSchemaManager.java:814)
    ... 59 more
2015-10-09 10:30:10,366 [main] WARN  org.springframework.web.context.support.GenericWebApplicationContext - Exception encountered during context initialization - cancelling refresh attempt
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManager' defined in class path resource [com/shaka/dataaccess/config/DataAccessConfiguration.class]: Invocation of init method failed; nested exception is com.impetus.kundera.configure.schema.SchemaGenerationException: com.impetus.kundera.configure.schema.SchemaGenerationException: Error occurred while creating table subject
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1578)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
    at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:305)
    at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
    at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:301)
    at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:196)
    at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1045)
    at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:824)
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:537)
    at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:687)
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
    at org.springframework.boot.test.SpringApplicationContextLoader.loadContext(SpringApplicationContextLoader.java:104)
    at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
    at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
    at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
    at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:183)
    at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:123)
    at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:228)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:217)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:276)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:278)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:236)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:87)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
    at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:180)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: com.impetus.kundera.configure.schema.SchemaGenerationException: com.impetus.kundera.configure.schema.SchemaGenerationException: Error occurred while creating table subject
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.update(CassandraSchemaManager.java:632)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.handleOperations(AbstractSchemaManager.java:267)
    at com.impetus.kundera.configure.schema.api.AbstractSchemaManager.exportSchema(AbstractSchemaManager.java:115)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.exportSchema(CassandraSchemaManager.java:166)
    at com.impetus.kundera.configure.SchemaConfiguration.configure(SchemaConfiguration.java:188)
    at com.impetus.kundera.configure.ClientMetadataBuilder.buildClientFactoryMetadata(ClientMetadataBuilder.java:48)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configureClientFactories(EntityManagerFactoryImpl.java:408)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configure(EntityManagerFactoryImpl.java:161)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.<init>(EntityManagerFactoryImpl.java:116)
    at com.impetus.kundera.KunderaPersistence.createContainerEntityManagerFactory(KunderaPersistence.java:63)
    at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:343)
    at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:318)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
    ... 43 more
Caused by: com.impetus.kundera.configure.schema.SchemaGenerationException: Error occurred while creating table subject
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.updateExistingColumnFamily(CassandraSchemaManager.java:573)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateUsingCQL3(CassandraSchemaManager.java:820)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createColumnFamilies(CassandraSchemaManager.java:430)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateKeyspace(CassandraSchemaManager.java:279)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.update(CassandraSchemaManager.java:627)
    ... 56 more
Caused by: InvalidRequestException(why:Cannot add already existing table "subject" to keyspace "shaka_test")
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result$execute_cql3_query_resultStandardScheme.read(Cassandra.java:50297)
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result$execute_cql3_query_resultStandardScheme.read(Cassandra.java:50274)
    at org.apache.cassandra.thrift.Cassandra$execute_cql3_query_result.read(Cassandra.java:50189)
    at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:78)
    at org.apache.cassandra.thrift.Cassandra$Client.recv_execute_cql3_query(Cassandra.java:1734)
    at org.apache.cassandra.thrift.Cassandra$Client.execute_cql3_query(Cassandra.java:1719)
    at com.impetus.client.cassandra.schemamanager.CassandraSchemaManager.createOrUpdateUsingCQL3(CassandraSchemaManager.java:814)
    ... 59 more
```

And when I did debug on Kundera sources, I've found the problem is what I've described above. Kundera expects the 'why' field of the InvalidRequestException to contain this "Cannot add already existing column family " but instead, the actual message says: "Cannot add already existing table ". Even when you fix the expected message, you're certainly relying on a string that is not under Kundera's control, hence the table existence detection engine is kind of weak in my opinion. That's why I've suggested to use CQL queries to check for table existence or perhaps this solution based on the Java Datastax driver:

```
KeyspaceMetadata ks = cluster.getMetadata().getKeyspace("myKeyspace");
TableMetadata table = ks.getTable("myTable");
```

**Comment by karthikprasad13** [comment]
@xshaka 

Issue occurs when using Kundera with Cassandra 2.2+ because Kundera is not yet 100% compatible with Cassandra 2.2+. We are taking this as a part of Kundera-Cassandra upgrade and will be resolved soon. This works with Cassandra 2.1+.

-Karthik

**Comment by chhavigangwal** [comment]
Fix released with Kundera-3.2.
